cpp/src/strings/split/split_re.cu

```diff
@@ -0,0 +1,375 @@
+  auto const regex_insts = d_prog.insts_counts();
+  if (regex_insts <= RX_SMALL_INSTS) {
+    token_reader_fn<RX_STACK_SMALL> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  } else if (regex_insts <= RX_MEDIUM_INSTS) {
+    token_reader_fn<RX_STACK_MEDIUM> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  } else if (regex_insts <= RX_LARGE_INSTS) {
+    token_reader_fn<RX_STACK_LARGE> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  } else {
+    token_reader_fn<RX_STACK_ANY> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  }
+
+  return tokens;
+}
+
+/**
+ * @brief Returns string pair for the specified column for each string in `d_strings`
+ *
+ * This is used to build the table result of a split.
+ * Null is returned if the row is null or if the `column_index` is larger
+ * than the token count for that string.
+ */
+struct tokens_transform_fn {
+  column_device_view const d_strings;
+  string_index_pair const* d_tokens;
+  offset_type const* d_token_offsets;
+  size_type const column_index;
+
+  __device__ string_index_pair operator()(size_type idx) const
+  {
+    auto const offset      = d_token_offsets[idx];
+    auto const token_count = d_token_offsets[idx + 1] - offset;
+    return (column_index >= token_count) || d_strings.is_null(idx)
+             ? string_index_pair{nullptr, 0}
+             : d_tokens[offset + column_index];
+  }
+};
+
+std::unique_ptr<table> split_re(strings_column_view const& input,
+                                std::string const& pattern,
+                                split_direction direction,
+                                size_type maxsplit,
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(!pattern.empty(), "Parameter pattern must not be empty");
+
+  auto const strings_count = input.size();
+
+  std::vector<std::unique_ptr<column>> results;
+  if (strings_count == 0) {
+    results.push_back(make_empty_column(type_id::STRING));
+    return std::make_unique<table>(std::move(results));
+  }
+
+  // create the regex device prog from the given pattern
+  auto d_prog = reprog_device::create(pattern, get_character_flags_table(), strings_count, stream);
+  auto d_strings = column_device_view::create(input.parent(), stream);
+
+  // count the number of delimiters matched in each string
+  auto offsets = count_matches(*d_strings, *d_prog, stream, rmm::mr::get_current_device_resource());
+  auto offsets_view = offsets->mutable_view();
+  auto d_offsets    = offsets_view.data<offset_type>();
```

Comment:
Would it make sense to roll this into `generate_tokens` and have it return the tokens/offsets as a pair?  It already has everything it needs to do the `count_matches` step and the other use case below (`split_record_re`) does the same thing.

Reply:
They offsets are not technically generated the same. In the `split_re` the offsets are only temporary working data but in the `split_record_re` they are to be returned. So each is built with a different memory resource as appropriate. Though seemingly minor, I feel this is an important distinction between the two implementations.